firmware/controllers/gauges/tachometer.cpp

```diff
@@ -18,26 +18,71 @@
 EXTERN_ENGINE;
 
 static scheduling_s tachTurnSignalOff;
+static float halfperiodMs = 4; /* period in ms, used for multi pulse output, actually half period */
+static uint8_t isStarted = 0;
+
+static void toggleTachPinHigh(void);
+static void toggleTachPinLow(void);
 
 static void turnTachPinLow(void) {
 	enginePins.tachOut.setLow();
 }
 
+static void toggleTachPinLow(void)
+{
+	if(isStarted)
+	{
+		enginePins.tachOut.setLow();
+		engine->executor.scheduleForLater(&tachTurnSignalOff, (int)MS2US(halfperiodMs), (schfunc_t) &toggleTachPinHigh, NULL);
+	}
+}
+
+static void toggleTachPinHigh(void)
+{
+	if(isStarted)
+	{
+		enginePins.tachOut.setHigh();
+		engine->executor.scheduleForLater(&tachTurnSignalOff, (int)MS2US(halfperiodMs), (schfunc_t) &toggleTachPinLow, NULL);
+	}
+
+}
+
 static void tachSignalCallback(trigger_event_e ckpSignalType,
 		uint32_t index DECLARE_ENGINE_PARAMETER_SUFFIX) {
 	UNUSED(ckpSignalType);
+
 	if (index != (uint32_t)engineConfiguration->tachPulseTriggerIndex) {
 		return;
 	}
-	enginePins.tachOut.setHigh();
-	float durationMs;
-	if (engineConfiguration->tachPulseDurationAsDutyCycle) {
-		// todo: implement tachPulseDurationAsDutyCycle
-		durationMs = engineConfiguration->tachPulseDuractionMs;
-	} else {
-		durationMs = engineConfiguration->tachPulseDuractionMs;
+
+	if ((1 < engineConfiguration->tachPulsePerRev))
+	{
+
+		/* reset timing counters */
+		int rpm = GET_RPM();
+		float duration = 60000/rpm;
```

Comment:
int divided by int is an int. That's scary that we have this variable defined as float :(

Reply:
If you want to actually end up with a float (which you do!), then you should do `60000.0 / rpm`

Reply:
Float was from the original implementation, then i missed to change it to int.

Just checked the method scheduleForLater, dose it make sense to use float? at the end it will be casted to int anyway.

Reply:
You actually do - even though `scheduleForLater` takes an integer, it takes an integer number of microseconds, not milliseconds.  When you call `MS2US` it multiplies by 1000.

For example, if the engine was running at 6001 RPM:
60000 / 6001 (integer) = 9
Which multiplied by 1000 gives 9000us.  A period of 9000us is actually 6666 RPM - an ~11% error.

Doing the intermediate work as a float with conversion to int at the end gives 9.998ms, or 9998us.  A period of 9998us is 6001.20 RPM, a 0.1% error

Reply:
You are right, checked!